SQL Queries and Transaction Handling

stino-x opened the pull request:
I hope this message finds you well. I am writing to submit a pull request that includes several SQL queries and transaction handling operations, based on the knowledge and guidance I've received today. These queries and operations are designed to extract useful information from the database and demonstrate transaction management. Here is an overview of what's included in this pull request:

SQL Queries:

A query to count the total number of animals in the database.
A query to count the animals that have never tried to escape.
A query to calculate the average weight of all animals.
A query to find which group (neutered or not neutered) has the highest number of escape attempts.
A query to determine the minimum and maximum weight for each species of animals.
A query to calculate the average number of escape attempts for animals born between 1990 and 2000, grouped by species.
Transaction Handling:

A transaction that deletes all records of animals born after January 1st, 2022, and then rolls back the transaction.
A transaction that updates all animals' weights to be their weight multiplied by -1, creates a savepoint within the transaction, rolls back to the savepoint, updates weights again for animals with negative weights, and finally commits the transaction.
These SQL queries and transaction operations aim to provide valuable insights into the database while showcasing the use of transactions for data integrity.

I would appreciate your review and feedback on this pull request. Please feel free to reach out if you have any questions or suggestions.

Thank you for your time and consideration.

Sincerely,
[IHEAGWARA AUGUSTINE]

*IMAGES*
PART-1: 
<img src= https://github.com/stino-x/vet-clinic-psql/assets/107426247/d4c99bb6-2c82-49ff-865f-516725e8f9d4>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/1fd16452-0d54-429b-baa3-adbe40a567bd>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/baf6848b-297d-4f94-897a-5ddfaf526bf3>


PART-2: 
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/b5beba6b-5c7e-4dd0-8cca-840a3e024395>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/dfde4308-323c-48c4-a811-8ed66195c991>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/8a41114d-b013-47eb-ad2e-8206dc10fe3b>

PART-3: 
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/03929f66-f062-4904-8c58-2366a93cfdb0>

PART-4: 
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/9db88bcf-c929-4783-b1df-039b7b5f4e10>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/c4630f63-6a73-4eba-8d49-ea6092924a73>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/86b62b05-ff21-444e-895b-d7ded627c870>
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/1b43848b-4eb7-495d-af33-7509257aa47c>

*ADDED IDS*
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/01f1a7fc-8fb2-4063-9652-fe1888819679 >

*attached proof of species step (making the species column unspecified)*
<img src=https://github.com/stino-x/vet-clinic-psql/assets/107426247/5883a166-dd10-44a5-b860-371a8c0c7324>

stino-x committed:
Task 2:

1. Added several animals with negative weights to the database.
2. Updated animals with names ending in "mon" to have species as "digimon".
3. Updated animals without a species to have species as "pokemon".
4. Deleted rows where the name is an empty string or NULL.
5. Started a transaction, deleted all records in the animals table, and rolled back the transaction.
6. Verified if all records in the animals table still exist.
7. Started a transaction, deleted all animals born after Jan 1st, 2022, created a savepoint, updated all animals' weight to be their weight multiplied by -1, rolled back to the savepoint, and updated all animals' weights that are negative to be their weight multiplied by -1.
8. Committed the transaction. (generated by useblackbox.ai )

skyv26 commented on data.sql:
```diff
@@ -16,3 +16,47 @@ VALUES ('Pikachu', '2021-01-07', 15.04, false, 1);
+---- TASK 2-------
+
+-- Inserting the first animal (Charmander) with a negative weight
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Charmander', '2020-02-08', -11.0, false, 0, 'Fire-type');
+
+-- Inserting the second animal (Plantmon) with a negative weight
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Plantmon', '2021-11-15', -5.7, true, 2, 'Grass-type');
+
+-- Inserting the third animal (Squirtle) with a negative weight
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Squirtle', '1993-04-02', -12.13, false, 3, 'Water-type');
+
+-- Inserting the fourth animal (Angemon) with a negative weight
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Angemon', '2005-06-12', -45.0, true, 1, 'Holy-type');
+
+-- Inserting the fifth animal (Boarmon)
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Boarmon', '2005-06-07', 20.4, true, 7, 'Ground-type');
+
+-- Inserting the sixth animal (Blossom)
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Blossom', '1998-10-13', 17.0, true, 3, 'Grass-type');
+
+-- Inserting the seventh animal (Ditto)
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Ditto', '2022-05-14', 22.0, true, 4, 'Normal-type');
```
- [ ] Please always double-check your work before submitting your project for review. I tried to see what you did and I got an error related to `id`, because this ID column is not null and in order to resolve the problem, a user must add the unique ID himself or herself. 

![image](https://github.com/stino-x/vet-clinic-psql/assets/66476329/e1a2191c-c612-4bef-a5fa-367ecaa056f8)

skyv26 commented on data.sql:
```diff
@@ -16,3 +16,47 @@ VALUES ('Pikachu', '2021-01-07', 15.04, false, 1);
+INSERT INTO animals (name, date_of_birth, weight_kg, neutered, escape_attempts, species)
+VALUES ('Ditto', '2022-05-14', 22.0, true, 4, 'Normal-type');
+
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 101448.png
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 101524.png
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 101556.png
+
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 103209.png
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 102931.png
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 103024.png
+
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 103520.png
+
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 104118.png
+
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 105353.png
+-- c:\Users\Danskids\Downloads\Screenshot 2023-09-30 105441.png
```
- [ ] Please remove these unnecessary comments from your working directory files, because they look unprofessional and also can put a person into confusion.

skyv26 commented on queries.sql:
- [ ] Kindly read the requirements carefully and also take a look at your work. I think you missed one more requirement, where you should add the `unspecified` text for the species columns as data. I looked into your code and I didn't find anything related to the requirement attached below.
![image](https://github.com/stino-x/vet-clinic-psql/assets/66476329/9256e070-56d9-4db3-b13f-f7e185ee75a9)

skyv26 reviewed (changes requested):
## Hi @stino-x, 

Good job so far 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## STATUS: REQUIRE CHANGES 🔴 

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYjBkZWI1YTgxNDg4Y2QzMWU2MjkyNjQ2MmY1MmIzMjM4ZmI1MGU1OCZlcD12MV9pbnRlcm5hbF9naWZzX2dpZklkJmN0PWc/fDWHJAmgrCeIw/giphy.gif)

## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implement the features ✅ 
- Added  the screen-shots ✅ 

## Required Changes ♻️

- [ ] Please 🙏🏻 adhere to the project requirement, according to which you should have to alter the table and add a `species` column in the `schema.sql` file. It seems like you didn't do the things as per the requirement. Kindly look at the below requirement and submit your project again for review.

### Requirement
![Screenshot from 2023-09-30 16-10-18](https://github.com/stino-x/vet-clinic-psql/assets/66476329/4b0b6c62-6968-4c10-b91c-445adbbef841)


- [ ] Please 🙏🏻  add the `MIT` license properly because including the **MIT** *License* in your `GitHub` repository **README** is crucial for several reasons. Firstly, the **MIT** License is a permissive `open-source` license that grants users the freedom to use, modify, and distribute the software without legal restrictions. By choosing the *MIT* License 🗒️, you are encouraging collaboration, inviting other developers to contribute, and fostering a vibrant open-source community. Additionally, using the *MIT* License helps protect your work by clearly defining the terms and conditions under which it can be used, providing a level of legal 🚨 clarity and ensuring that proper attribution is given. Furthermore, the widespread popularity and familiarity of the *MIT* License make it easy for others to understand and comply with the licensing requirements, which ultimately promotes the adoption and usage of your software 💻 .

- [ ] Please remove the irrelevant information from your `tech-stack` in the `README` file, because right now you mentioned that you are using `React` as a client and `Express` as the server but you using neither of them. Kindly remove these files.

![image](https://github.com/stino-x/vet-clinic-psql/assets/66476329/b9ee70e1-2cb8-4c80-94d3-db08e392f736)

- [ ] Please remove the unnecessary things from your `README` file, like the `getting-started` section before the `table of contents` and `README` comments. Comments are meant for developer guidance purpose, leaving the comment create a kind of confusion for a developer or user like me.

![Screenshot from 2023-09-30 16-45-24](https://github.com/stino-x/vet-clinic-psql/assets/66476329/199ea039-c76b-4438-b62f-762f4192d38f)

![Screenshot from 2023-09-30 16-45-04](https://github.com/stino-x/vet-clinic-psql/assets/66476329/5a7f71a7-8f0a-4d8f-b2e6-d2df68aa2a4e)

![Screenshot from 2023-09-30 16-44-54](https://github.com/stino-x/vet-clinic-psql/assets/66476329/079f78a3-7149-4ef3-8fc1-17a891564170)


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

#### Cheers, and Happy coding!👏👏👏

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   <img src="https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white" alt="@skyv26" />
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   <img src="https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white" alt="linkedin-skyv2022" />
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   <img src="https://img.shields.io/badge/Twitter-1DA1F2?style=for-the-badge&logo=twitter&logoColor=white" alt="@vrma_aakash" />
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   <img src="https://img.shields.io/badge/Instagram-E4405F?style=for-the-badge&logo=instagram&logoColor=white" alt="@skybrel" />
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   <img src="https://user-images.githubusercontent.com/66476329/243158690-61eb1021-fbb6-464d-b841-a09c4caa156e.png" alt="@skybrel" />
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
Update data.sql and queries.sql
 i also included the step where they said we should make the species unspecified as i did it already i just forgot to attach the image earlier

Alusp commented on data.sql:
```diff
@@ -16,3 +16,35 @@ VALUES ('Pikachu', '2021-01-07', 15.04, false, 1);
 VALUES ('Devimon', '2017-05-12', 11.0, true, 5);
 
+---- TASK 2-------
+UPDATE animals
+SET species = 'unspecified';
+UPDATE 7
```
- [ ] Good job implementing the changes requested by the previous reviewer, :clap: as he mentioned Please ensure that your queries related to applying operations like `update`, `delete` etc should be inside the `queries.sql` not in the `data.sql` file. As I can see you added the same update command in the `queries.sql` too but still you missed the requirements of adding the command under the transactions. Which means you missed both of the requirements as I said. Kindly go through the requirements, and read the text carefully. :slightly_smiling_face:

Alusp commented on queries.sql:
```diff
@@ -24,4 +24,93 @@ SELECT * FROM animals WHERE name <> 'Gabumon';
 SELECT * FROM animals WHERE weight_kg >= 10.4 AND weight_kg <= 17.3;
 
+--- Task 2 ----
+UPDATE animals
+SET species = 'unspecified';
```
- [ ] This is the line of code I am referring to: Please put it below `Begin Transaction` as a requirement.

Alusp reviewed (changes requested):
Hi @stino-x ,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
 ![close](https://media.giphy.com/media/nJKDTpuxnZE7fb6Z1b/giphy.gif)

 ## Highlights

- [ ] Good PR title ✅
- [ ] Descriptive PR description ✅
- [ ] Implement the features ✅

 ## Required Changes ♻️

- [ ] Good job so far, :clap: However, as mentioned the previewer, Please remove the Tech Stack Header that says `REACT` and `Express` and also ensure to remove all comments in your Readme File. :slightly_smiling_face: 

_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

stino-x committed:
The diff contains changes to two files: data.sql and queries.sql.

In data.sql, the following changes were made:
-  The value for the 'species' column of the row with the name 'Devimon' was changed from 'unspecified' to 'digimon'.
- The 'species' column was removed from the INSERT statement for the row with the name 'Charmander'.

In queries.sql, the following changes were made:
-  The UPDATE statement that sets the 'species' column to 'unspecified' for all rows was removed.
- A COMMIT statement was added. (generated by useblackbox.ai )

IngDavidTM reviewed (approved):
Hi @stino-x 🙋‍♂️, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request follows the GitFlow ✅ 
- The PR title and description are professional and descriptive ✅ 
- Linters are working properly ✅ 
<img src="https://media3.giphy.com/media/xT0xezQGU5xCDJuCPe/giphy.gif"/>

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
